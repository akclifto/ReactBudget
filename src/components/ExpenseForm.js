import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

// date object for react sucks, using moment.js instead. Documentation is really good for moment
//const date = new Date();
// const now = moment();
// console.log(now.format('MMM DD YYYY'));

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props)

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    setupNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;

        // using a regular expression here to get a correct value and char input for money to two decimal places
        // search regex101 to learn more about it
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }

    onDateChange = (createdAt) => {

        if(createdAt) {
            this.setState(() => ( { createdAt }));
        }

    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ( { calendarFocused: focused }))
    }

    onSubmit = (e) => {
        
        e.preventDefault();
        if(!this.state.description || !this.state.amount) {
            // set error state 
            this.setState(() => ( { error: 'Please provide description and amount!' }));
        } else {
            // clear the error
            this.setState(() => ( { error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
            // console.log('submitted');
        }
    };

    render() {
        return (

            <div>
            {this.state.error && <p>{this.state.error}</p>}

                <form onSubmit={this.onSubmit}>

                    <input
                        type="text"
                        placeholder="Description of Expense"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />

                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        //number of calendar months visible when selecting a day
                        numberOfMonths={1}
                        //inline way to let you select any day in existence 
                        isOutsideRange={ () => false}
                    />
                    
                    <br />

                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.setupNoteChange}
                    >
                    </textarea>

                    <br />

                    <button>Save Expense</button>
                </form>
            </div>
        )
    }
}
