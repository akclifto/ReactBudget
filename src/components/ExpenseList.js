import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListitem';
import selectExpenses from '../selectors/get-expenses';


export const ExpenseList = (props) => (

    <div>
        {
            props.expenses.length === 0 ? (
                <p> No Expenses in List </p>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense} />
                })
            )
        }
    </div>

);


const mapStateToProps = (state) => {

    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

// We can use connect from redux and the state of the components to access information as needed.  
// The connect components lets us access info within other components where the structure may be disjointed
//      meaning, there is no direct link between two components that need to connect information.  
//      this apparently is pretty common.

//  connect (the state of the props to be mapped)(where the connect will be mapped)

export default connect(mapStateToProps)(ExpenseList);
