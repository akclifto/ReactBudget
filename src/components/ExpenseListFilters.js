import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortbyDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { v4 as uuidv4 } from 'uuid';

//make a text filter to search or find things.  
export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
        // console.log(e.target.value)
    };

    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortbyDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    };


    render() {
        return (

            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />

                {/*Below is how to create a controlled input using a text filtering process and drop down selectors.*/}

                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>

                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId={uuidv4()}
                    endDate={this.props.filters.endDate}
                    endDateId={uuidv4()}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />

            </div>
        );
    }

}

const mapDispatchToProps = (dispatch) => ({

    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setStartDate(endDate)),
    setTextFilter: (text) => dispatch(setStartDate(text)),
    sortbyDate: () => dispatch(setStartDate()),
    sortByAmount: () => dispatch(setStartDate())
});

const mapStateToProps = (state) => {

    return {
        filters: state.filters
    };

};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);