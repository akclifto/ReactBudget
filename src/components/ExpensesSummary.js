import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/get-expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenseCount, expenseTotal}) => {

    const expenseWord = expenseCount === 1 ? 'expense': 'expenses';
    const expenseTotalFormatted = numeral(expenseTotal / 100 ).format('$0,0.00');

    return (
        <div>
            <h1> Viewing {expenseCount} {expenseWord} totalling {expenseTotalFormatted}</h1>
        </div>
    )

};

// need to map state to props for connector
const mapStateToProps = (state) => {

    const visibleExpenses = selectExpenses(state.expenses, state.filters);

    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: selectExpensesTotal(visibleExpenses)
    };
};

//connector react-redux
export default connect(mapStateToProps)(ExpensesSummary);