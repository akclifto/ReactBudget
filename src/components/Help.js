import React from 'react';


const Help = () => (
    <div>
        This is from the help page component!
    </div>

);

export default Help;
