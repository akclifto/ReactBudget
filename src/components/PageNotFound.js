import React from 'react';
import { Link } from 'react-router-dom';


const PageNotFound = () => (
    <div>
        <h1>404! NOT FOUND PAGE</h1>
        <Link to="/">Go Home</Link>

    </div>

);

export default PageNotFound;
