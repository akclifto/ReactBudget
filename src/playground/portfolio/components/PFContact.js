import React from 'react';


const PFContact = () => (
    <div>
        <h1>Contact Me</h1>
        This is where you contact me at test@gmail.com!
    </div>

);

export default PFContact ;