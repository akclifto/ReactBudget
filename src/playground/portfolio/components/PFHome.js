import React from 'react';


const PFHome = () => (

    <div>
        <h1>Welcome</h1>
        This is my site.  Take a look around!
    </div>
);

export default PFHome;