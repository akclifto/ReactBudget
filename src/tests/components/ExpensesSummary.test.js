import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('Should render the Expenses Summary Correctly with 1 expense (snapshot)', () => {
    const wrapper = shallow(<ExpensesSummary 
        expenseCount={1} 
        expenseTotal={195}
        />)

        expect(wrapper).toMatchSnapshot();
});

test('Should render the Expenses Summary Correctly with multiple expense (snapshot)', () => {
    const wrapper = shallow(<ExpensesSummary 
        expenseCount={13} 
        expenseTotal={2044541}
        />)

        expect(wrapper).toMatchSnapshot();
});

test(' ', () => {

});