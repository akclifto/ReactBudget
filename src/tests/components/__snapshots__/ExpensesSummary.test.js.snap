// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the Expenses Summary Correctly with 1 expense (snapshot) 1`] = `
<div>
  <h1>
     Viewing 
    1
     
    expense
     totalling 
    $1.95
  </h1>
</div>
`;

exports[`Should render the Expenses Summary Correctly with multiple expense (snapshot) 1`] = `
<div>
  <h1>
     Viewing 
    13
     
    expenses
     totalling 
    $20,445.41
  </h1>
</div>
`;
